<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Introduction to nf-core &mdash; Courses@CRG: Reproducible research and data analysis using Containers &amp; Nextflow 2022  documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
    <link rel="canonical" href="https://biocorecrg.github.io/CoursesCRG_Containers_Nextflow_May_2022/nf-core_1.html" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="_static/doctools.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="author" title="About these documents" href="about.html" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="nf-core for developers" href="nf-core_2.html" />
    <link rel="prev" title="Nextflow 3" href="nextflow_3.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="index.html" class="icon icon-home"> Courses@CRG: Reproducible research and data analysis using Containers & Nextflow 2022
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="about.html">About the course</a></li>
<li class="toctree-l1"><a class="reference internal" href="containers.html">Containers</a></li>
<li class="toctree-l1"><a class="reference internal" href="docker.html">Docker</a></li>
<li class="toctree-l1"><a class="reference internal" href="docker_2.html">Docker 2</a></li>
<li class="toctree-l1"><a class="reference internal" href="singularity.html">Singularity</a></li>
<li class="toctree-l1"><a class="reference internal" href="nextflow_1.html">Nextflow 1</a></li>
<li class="toctree-l1"><a class="reference internal" href="nextflow_2.html">Nextflow 2</a></li>
<li class="toctree-l1"><a class="reference internal" href="nextflow_3.html">Nextflow 3</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Introduction to nf-core</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#community">Community</a></li>
<li class="toctree-l2"><a class="reference internal" href="#pipelines">Pipelines</a></li>
<li class="toctree-l2"><a class="reference internal" href="#standard">Standard</a></li>
<li class="toctree-l2"><a class="reference internal" href="#helper-tools">Helper tools</a></li>
<li class="toctree-l2"><a class="reference internal" href="#installation">Installation</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="#nf-core-for-users">nf-core for users</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#listing-pipelines">Listing pipelines</a></li>
<li class="toctree-l2"><a class="reference internal" href="#filtering-available-nf-core-pipelines">Filtering available nf-core pipelines</a></li>
<li class="toctree-l2"><a class="reference internal" href="#pulling-pipelines">Pulling pipelines</a></li>
<li class="toctree-l2"><a class="reference internal" href="#launching-pipelines">Launching pipelines</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#exercise">Exercise</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#nf-core-configs-and-profiles">nf-core configs and profiles</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#nf-core-configs">nf-core configs</a></li>
<li class="toctree-l3"><a class="reference internal" href="#nf-core-core-profiles">nf-core core profiles</a></li>
<li class="toctree-l3"><a class="reference internal" href="#institutional-profiles">Institutional profiles</a></li>
<li class="toctree-l3"><a class="reference internal" href="#custom-config">Custom config</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id1">Exercise</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#running-pipelines-with-test-data">Running pipelines with test data</a></li>
<li class="toctree-l2"><a class="reference internal" href="#downloading-pipelines">Downloading pipelines</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="nf-core_2.html">nf-core for developers</a></li>
<li class="toctree-l1"><a class="reference internal" href="nf-core_2.html#nf-core-modules">nf-core modules</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">Courses@CRG: Reproducible research and data analysis using Containers & Nextflow 2022</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
      <li>Introduction to nf-core</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/nf-core_1.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <a class="reference internal image-reference" href="_images/nf-core-logo.png" id="nf-core-1-page"><img alt="_images/nf-core-logo.png" id="nf-core-1-page" src="_images/nf-core-logo.png" style="width: 400px;" /></a>
<section id="introduction-to-nf-core">
<h1>Introduction to nf-core<a class="headerlink" href="#introduction-to-nf-core" title="Permalink to this heading"></a></h1>
<p>nf-core is a community effort to collect a curated set of analysis pipelines built using Nextflow.</p>
<section id="community">
<h2>Community<a class="headerlink" href="#community" title="Permalink to this heading"></a></h2>
</section>
<section id="pipelines">
<h2>Pipelines<a class="headerlink" href="#pipelines" title="Permalink to this heading"></a></h2>
</section>
<section id="standard">
<h2>Standard<a class="headerlink" href="#standard" title="Permalink to this heading"></a></h2>
</section>
<section id="helper-tools">
<h2>Helper tools<a class="headerlink" href="#helper-tools" title="Permalink to this heading"></a></h2>
<p>The main nf-core was published in 2020 in <a class="reference external" href="https://pubmed.ncbi.nlm.nih.gov/32055031/">Nature Biotechnology</a> and
describes the community and framework.</p>
<a class="reference internal image-reference" href="_images/nf-core-paper.png"><img alt="_images/nf-core-paper.png" src="_images/nf-core-paper.png" style="width: 600px;" /></a>
</section>
<section id="installation">
<h2>Installation<a class="headerlink" href="#installation" title="Permalink to this heading"></a></h2>
<p>You can use Conda to install nf-core tools, in the command below we create a new named environment that includes nf-core
and then, we activate it.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">conda create --name nf-core nf-core=2.4.1 -c bioconda -c conda-forge -y</span>
<span class="go">conda activate nf-core</span>
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>You will need to install again Nextflow in the instances using <code class="docutils literal notranslate"><span class="pre">curl</span> <span class="pre">-s</span> <span class="pre">https://get.nextflow.io</span> <span class="pre">|</span> <span class="pre">bash</span></code> and move the
executable to <code class="docutils literal notranslate"><span class="pre">sudo</span> <span class="pre">mv</span> <span class="pre">nextflow</span> <span class="pre">/usr/local/bin</span></code></p>
</div>
<div class="admonition tip">
<p class="admonition-title">Tip</p>
<p>Find alternative ways of installation on the nf-core <a class="reference external" href="https://nf-co.re/tools/#installation">documentation</a></p>
</div>
<p>We can now check nf-core available commands:</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">    $ </span>nf-core
<span class="go">                                      ,--./,-.</span>
<span class="go">      ___     __   __   __   ___     /,-._.--~\</span>
<span class="go">|\ | |__  __ /  ` /  \ |__) |__         }  {</span>
<span class="go">| \| |       \__, \__/ |  \ |___     \`-._,-`-,</span>
<span class="go">                                      `._,._,&#39;</span>

<span class="go">nf-core/tools version 2.4.1 - https://nf-co.re</span>



<span class="go">    Usage: nf-core [OPTIONS] COMMAND [ARGS]...</span>

<span class="go">    nf-core/tools provides a set of helper tools for use with nf-core Nextflow pipelines.</span>
<span class="go">    It is designed for both end-users running pipelines and also developers creating new pipelines.</span>

<span class="go">    ╭─ Options ────────────────────────────────────────────────────────────────────────────────────────╮</span>
<span class="go">    │  --version                   Show the version and exit.                                          │</span>
<span class="go">    │  --verbose   -v              Print verbose output to the console.                                │</span>
<span class="go">    │  --log-file  -l  &lt;filename&gt;  Save a verbose log to a file.                                       │</span>
<span class="go">    │  --help      -h              Show this message and exit.                                         │</span>
<span class="go">    ╰──────────────────────────────────────────────────────────────────────────────────────────────────╯</span>
<span class="go">    ╭─ Commands for users ─────────────────────────────────────────────────────────────────────────────╮</span>
<span class="go">    │  list      List available nf-core pipelines with local info.                                     │</span>
<span class="go">    │  launch    Launch a pipeline using a web GUI or command line prompts.                            │</span>
<span class="go">    │  download  Download a pipeline, nf-core/configs and pipeline singularity images.                 │</span>
<span class="go">    │  licences  List software licences for a given workflow.                                          │</span>
<span class="go">    ╰──────────────────────────────────────────────────────────────────────────────────────────────────╯</span>
<span class="go">    ╭─ Commands for developers ────────────────────────────────────────────────────────────────────────╮</span>
<span class="go">    │  create        Create a new pipeline using the nf-core template.                                 │</span>
<span class="go">    │  lint          Check pipeline code against nf-core guidelines.                                   │</span>
<span class="go">    │  modules       Commands to manage Nextflow DSL2 modules (tool wrappers).                         │</span>
<span class="go">    │  schema        Suite of tools for developers to manage pipeline schema.                          │</span>
<span class="go">    │  bump-version  Update nf-core pipeline version number.                                           │</span>
<span class="go">    │  sync          Sync a pipeline TEMPLATE branch with the nf-core template.                        │</span>
<span class="go">    ╰──────────────────────────────────────────────────────────────────────────────────────────────────╯</span>
</pre></div>
</div>
<p>As shown in the screenshot, nf-core tools provide with some commands meant for users and with some commands meant
for developers. We will start first discussing how nf-core can be used from a user point of view.</p>
</section>
</section>
<section id="nf-core-for-users">
<h1>nf-core for users<a class="headerlink" href="#nf-core-for-users" title="Permalink to this heading"></a></h1>
<section id="listing-pipelines">
<h2>Listing pipelines<a class="headerlink" href="#listing-pipelines" title="Permalink to this heading"></a></h2>
<p>To show all the nf-core available pipelines, we can use the nf-core list command. This command also
provides some other information as the last version of each of the nf-core pipelines, its publication and
and when you last pulled the pipeline to your local system.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">    $ </span>nf-core list

<span class="go">                                      ,--./,-.</span>
<span class="go">      ___     __   __   __   ___     /,-._.--~\</span>
<span class="go">|\ | |__  __ /  ` /  \ |__) |__         }  {</span>
<span class="go">| \| |       \__, \__/ |  \ |___     \`-._,-`-,</span>
<span class="go">                                      `._,._,&#39;</span>

<span class="go">nf-core/tools version 2.4.1 - https://nf-co.re</span>


<span class="go">    ┏━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━┳━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━┓</span>
<span class="go">    ┃ Pipeline Name          ┃ Stars ┃ Latest Release ┃      Released ┃  Last Pulled ┃ Have latest release? ┃</span>
<span class="go">    ┡━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━╇━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━┩</span>
<span class="go">    │ airrflow               │    19 │          2.2.0 │    2 days ago │            - │ -                    │</span>
<span class="go">    │ circdna                │     4 │          1.0.0 │    5 days ago │            - │ -                    │</span>
<span class="go">    │ smrnaseq               │    39 │          2.0.0 │    5 days ago │ 7 months ago │ No (v1.1.0)          │</span>
<span class="go">    │ rnaseq                 │   465 │          3.8.1 │   1 weeks ago │   6 days ago │ Yes (v3.8.1)         │</span>
<span class="go">    │ ampliseq               │    88 │          2.3.2 │   1 weeks ago │            - │ -                    │</span>
<span class="go">    │ mnaseseq               │     7 │          1.0.0 │   2 weeks ago │            - │ -                    │</span>
<span class="go">    │ rnafusion              │    71 │          2.0.0 │   3 weeks ago │            - │ -                    │</span>
<span class="go">    │ fetchngs               │    58 │            1.6 │   3 weeks ago │  2 weeks ago │ No (v1.5)            │</span>
<span class="go">    │ atacseq                │   107 │          1.2.2 │   3 weeks ago │  2 weeks ago │ Yes (v1.2.2)         │</span>
<span class="go">    │ mhcquant               │    19 │          2.3.1 │   4 weeks ago │            - │ -                    │</span>
<span class="go">    │ hicar                  │     2 │          1.0.0 │  1 months ago │            - │ -                    │</span>
<span class="go">    │ quantms                │     1 │            1.0 │  1 months ago │            - │ -                    │</span>
<span class="go">    │ eager                  │    66 │          2.4.4 │  2 months ago │            - │ -                    │</span>
<span class="go">    │ viralrecon             │    70 │          2.4.1 │  3 months ago │ 7 months ago │ No (v2.2)            │</span>
<span class="go">    │ cutandrun              │    27 │            1.1 │  5 months ago │            - │ -                    │</span>
<span class="go">    │ epitopeprediction      │    18 │          2.0.0 │  6 months ago │            - │ -                    │</span>
<span class="go">    │ nanoseq                │    76 │          2.0.1 │  6 months ago │            - │ -                    │</span>
<span class="go">    │ mag                    │    88 │          2.1.1 │  6 months ago │            - │ -                    │</span>
<span class="go">    │ bacass                 │    35 │          2.0.0 │  9 months ago │            - │ -                    │</span>
<span class="go">    │ bactmap                │    29 │          1.0.0 │ 12 months ago │            - │ -                    │</span>
<span class="go">    │ sarek                  │   168 │          2.7.1 │ 12 months ago │ 2 months ago │ Yes (v2.7.1)         │</span>
<span class="go">    [..truncated..]</span>
</pre></div>
</div>
<div class="admonition tip">
<p class="admonition-title">Tip</p>
<p>The pipelines can be sorted by latest release (<code class="docutils literal notranslate"><span class="pre">-s</span> <span class="pre">release</span></code>, default), by the last time you pulled a local copy
(<code class="docutils literal notranslate"><span class="pre">-s</span> <span class="pre">pulled</span></code>), alphabetically (<code class="docutils literal notranslate"><span class="pre">-s</span> <span class="pre">name</span></code>) or by the number of GitHub stars (<code class="docutils literal notranslate"><span class="pre">-s</span> <span class="pre">stars</span></code>).</p>
</div>
</section>
<section id="filtering-available-nf-core-pipelines">
<h2>Filtering available nf-core pipelines<a class="headerlink" href="#filtering-available-nf-core-pipelines" title="Permalink to this heading"></a></h2>
<p>It is also possible to use keywords after the <code class="docutils literal notranslate"><span class="pre">list</span></code> command so that the list of pipelines is shortened to those
matching the keywords or including them in the description. We can use the command below to filter on the <strong>rna</strong>
and <strong>rna-seq</strong> keywords:</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">    $ </span>nf-core list rna rna-seq

<span class="go">                                      ,--./,-.</span>
<span class="go">      ___     __   __   __   ___     /,-._.--~\</span>
<span class="go">|\ | |__  __ /  ` /  \ |__) |__         }  {</span>
<span class="go">| \| |       \__, \__/ |  \ |___     \`-._,-`-,</span>
<span class="go">                                      `._,._,&#39;</span>

<span class="go">nf-core/tools version 2.4.1 - https://nf-co.re</span>

<span class="go">    ┏━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━┳━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━┳━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━┓</span>
<span class="go">    ┃ Pipeline Name          ┃ Stars ┃ Latest Release ┃    Released ┃  Last Pulled ┃ Have latest release? ┃</span>
<span class="go">    ┡━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━╇━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━╇━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━┩</span>
<span class="go">    │ smrnaseq               │    39 │          2.0.0 │  5 days ago │ 7 months ago │ No (v1.1.0)          │</span>
<span class="go">    │ rnaseq                 │   465 │          3.8.1 │ 1 weeks ago │   6 days ago │ Yes (v3.8.1)         │</span>
<span class="go">    │ rnafusion              │    71 │          2.0.0 │ 3 weeks ago │            - │ -                    │</span>
<span class="go">    │ dualrnaseq             │     7 │          1.0.0 │ 1 years ago │            - │ -                    │</span>
<span class="go">    │ circrna                │    18 │            dev │           - │            - │ -                    │</span>
<span class="go">    │ lncpipe                │    23 │            dev │           - │            - │ -                    │</span>
<span class="go">    │ scflow                 │    12 │            dev │           - │            - │ -                    │</span>
<span class="go">    │ spatialtranscriptomics │     3 │            dev │           - │            - │ -                    │</span>
<span class="go">    └────────────────────────┴───────┴────────────────┴─────────────┴──────────────┴──────────────────────┘</span>
</pre></div>
</div>
</section>
<section id="pulling-pipelines">
<h2>Pulling pipelines<a class="headerlink" href="#pulling-pipelines" title="Permalink to this heading"></a></h2>
<p>Once we have identified the nf-core pipeline we want to use we can pull it using the Nextflow <a class="reference external" href="https://www.nextflow.io/docs/latest/sharing.html#pulling-or-updating-a-project">built-in functionality</a>.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">nextflow pull nf-core/&lt;PIPELINE&gt;</span>
</pre></div>
</div>
<div class="admonition tip">
<p class="admonition-title">Tip</p>
<p>Nextflow will also automatically pull a project if you use <code class="docutils literal notranslate"><span class="pre">nextflow</span> <span class="pre">run</span> <span class="pre">nf-core/&lt;PIPELINE&gt;</span></code></p>
</div>
</section>
<section id="launching-pipelines">
<h2>Launching pipelines<a class="headerlink" href="#launching-pipelines" title="Permalink to this heading"></a></h2>
<p>The <code class="docutils literal notranslate"><span class="pre">launch</span></code> command enables to launch nf-core, and also Nextflow, pipelines via a web-based graphical interface or an
interactive command-line wizard tool. This command becomes handy for pipelines with a considerable number of parameters
since it displays the documentation alongside each of the parameters and validate your inputs.</p>
<p>We can now launch an nf-core pipeline:</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">nf-core launch</span>
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The pipelines can be sorted by latest release (<code class="docutils literal notranslate"><span class="pre">-s</span> <span class="pre">release</span></code>, default), by the last time you pulled a local copy
(<code class="docutils literal notranslate"><span class="pre">-s</span> <span class="pre">pulled</span></code>), alphabetically (<code class="docutils literal notranslate"><span class="pre">-s</span> <span class="pre">name</span></code>) or by the number of GitHub stars (<code class="docutils literal notranslate"><span class="pre">-s</span> <span class="pre">stars</span></code>).</p>
</div>
<p>To render the description of the parameters, its grouping and defaults, the tool uses the <code class="docutils literal notranslate"><span class="pre">nextflow_schema.json</span></code>. This
JSON file is bundled with the pipeline and includes all the information mentioned above, see an example <a class="reference external" href="https://github.com/nf-core/rnaseq/blob/03d17893618c44075e4c91d83dc0e72b58f6f0f7/nextflow_schema.json">here</a>.</p>
<p>The chosen not default parameters are dumped into a JSON file called <code class="docutils literal notranslate"><span class="pre">nf-params.json</span></code>. This file can be provided to new
executions using the <code class="docutils literal notranslate"><span class="pre">-params-file</span></code> flag. See below an example:</p>
<div class="highlight-json notranslate"><div class="highlight"><pre><span></span><span class="p">{</span><span class="w"></span>
<span class="w">    </span><span class="nt">&quot;outdir&quot;</span><span class="p">:</span><span class="w"> </span><span class="s2">&quot;results&quot;</span><span class="p">,</span><span class="w"></span>
<span class="w">    </span><span class="nt">&quot;input&quot;</span><span class="p">:</span><span class="w"> </span><span class="s2">&quot;https://raw.githubusercontent.com/nf-core/test-datasets/rnaseq/samplesheet/v3.4/samplesheet_test.csv&quot;</span><span class="w"></span>
<span class="p">}</span><span class="w"></span>
</pre></div>
</div>
<p>It is a good practice in terms of reproducibility to explicitly indicate the version (revision) of the pipeline that
you want to use, this can be indicated using the <cite>-r</cite> flag e.g. <code class="docutils literal notranslate"><span class="pre">nf-core</span> <span class="pre">launch</span> <span class="pre">rnaseq</span> <span class="pre">-r</span> <span class="pre">3.8.1</span></code>.</p>
<section id="exercise">
<h3>Exercise<a class="headerlink" href="#exercise" title="Permalink to this heading"></a></h3>
<p>Pull version <code class="docutils literal notranslate"><span class="pre">3.8.1</span></code> of the nf-core/rnaseq pipeline, run it using the <code class="docutils literal notranslate"><span class="pre">nf-core</span> <span class="pre">launch</span></code> command and produce the <code class="docutils literal notranslate"><span class="pre">nf-params.json</span></code>.</p>
<details>
<summary><a>Solution</a></summary><div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">nextflow pull nf-core/rnaseq -r 3.8.1</span>
<span class="go">nf-core launch rnaseq -r 3.8.1</span>
</pre></div>
</div>
</section>
</section>
<section id="nf-core-configs-and-profiles">
<h2>nf-core configs and profiles<a class="headerlink" href="#nf-core-configs-and-profiles" title="Permalink to this heading"></a></h2>
<section id="nf-core-configs">
<h3>nf-core configs<a class="headerlink" href="#nf-core-configs" title="Permalink to this heading"></a></h3>
<p>We have already introduced Nextflow configuration files and profiles during the course. Config files are used by
nf-core pipelines to specify the computational requirements of the pipeline, define custom parameters and set which
software management system to be used (Docker, Singularity or Conda). As an example take a look to the <a class="reference external" href="https://github.com/nf-core/smrnaseq/blob/master/conf/base.config">base.config</a>
that is used to set sensible defaults for the computational resources needed by the pipeline.</p>
</section>
<section id="nf-core-core-profiles">
<h3>nf-core core profiles<a class="headerlink" href="#nf-core-core-profiles" title="Permalink to this heading"></a></h3>
<p>nf-core pipelines use profiles to bundle a set of configuration attributes. By doing so, we can activate these
attributes by using the <code class="docutils literal notranslate"><span class="pre">-profile</span></code> Nextflow command line option.  All nf-core pipelines come along with a set of common
“Core profiles” that include the <code class="docutils literal notranslate"><span class="pre">conda</span></code>, <code class="docutils literal notranslate"><span class="pre">docker</span></code> and <code class="docutils literal notranslate"><span class="pre">singularity</span></code> that define which software manager to use and the
<code class="docutils literal notranslate"><span class="pre">test</span></code> profile that specifies a minimal test dataset to check that the pipelines works properly.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Each configuration file can include one or several profiles</p>
</div>
</section>
<section id="institutional-profiles">
<h3>Institutional profiles<a class="headerlink" href="#institutional-profiles" title="Permalink to this heading"></a></h3>
<p>Institutional profiles are profiles where you can specify the configuration attributes for your institution system. They are
hosted in <a class="reference external" href="https://github.com/nf-core/configs">https://github.com/nf-core/configs</a> and all pipelines pull this repository when a pipeline is run. The idea is that
these profiles set the custom config attributes to run nf-core pipelines in your institution (scheduler, container technology,
resources, etc.). This way all the users in a cluster can make use of the profile just setting the profile of your institution
(<code class="docutils literal notranslate"><span class="pre">-profile</span> <span class="pre">institution</span></code>).</p>
<div class="admonition tip">
<p class="admonition-title">Tip</p>
<p>You can use more than profile at a time by separating them by a comma without space, e.g. <code class="docutils literal notranslate"><span class="pre">-profile</span> <span class="pre">test,docker</span></code></p>
</div>
</section>
<section id="custom-config">
<h3>Custom config<a class="headerlink" href="#custom-config" title="Permalink to this heading"></a></h3>
<p>If you need to provide any custom parameter or setting when running a nf-core pipeline, you can do it by creating a local custom
config file and add it to your command with the <code class="docutils literal notranslate"><span class="pre">-c</span></code> flag.</p>
<a class="reference internal image-reference" href="_images/nfcore_config.png"><img alt="_images/nfcore_config.png" src="_images/nfcore_config.png" style="width: 600px;" /></a>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Profiles will be prioritized from left to right in case conflicting settings are found.</p>
</div>
</section>
<section id="id1">
<h3>Exercise<a class="headerlink" href="#id1" title="Permalink to this heading"></a></h3>
<ul class="simple">
<li><p>1 - Create a custom config that sets <code class="docutils literal notranslate"><span class="pre">params.email</span></code> to your email address and try to run the pipeline.</p></li>
</ul>
<details>
<summary><a>Solution</a></summary><div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">params.email = &quot;youremail@address.com&quot;</span>
</pre></div>
</div>
</details><div class="line-block">
<div class="line"><br /></div>
<div class="line"><br /></div>
</div>
</section>
</section>
<section id="running-pipelines-with-test-data">
<h2>Running pipelines with test data<a class="headerlink" href="#running-pipelines-with-test-data" title="Permalink to this heading"></a></h2>
<p>All nf-core pipelines include a special configuration named <code class="docutils literal notranslate"><span class="pre">test</span></code>. This configuration defines all the files and parameters to test
all pipeline functionality with a minimal dataset. Thus, although the functionality of the pipeline is maintained often the results
are not meaningful. As an example, find on the snippet below the test configuration of the <cite>nf-core/rnaseq &lt;https://github.com/nf-core/rnaseq/&gt;`__</cite>
pipeline.</p>
<div class="highlight-groovy notranslate"><div class="highlight"><pre><span></span><span class="cm">/*</span>
<span class="cm">~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~</span>
<span class="cm">    Nextflow config file for running minimal tests</span>
<span class="cm">~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~</span>
<span class="cm">    Defines input files and everything required to run a fast and simple pipeline test.</span>
<span class="cm">    Use as follows:</span>
<span class="cm">        nextflow run nf-core/rnaseq -profile test,&lt;docker/singularity&gt; --outdir &lt;OUTDIR&gt;</span>
<span class="cm">----------------------------------------------------------------------------------------</span>
<span class="cm">*/</span><span class="w"></span>

<span class="n">params</span><span class="w"> </span><span class="o">{</span><span class="w"></span>
<span class="w">    </span><span class="n">config_profile_name</span><span class="w">        </span><span class="o">=</span><span class="w"> </span><span class="s1">&#39;Test profile&#39;</span><span class="w"></span>
<span class="w">    </span><span class="n">config_profile_description</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s1">&#39;Minimal test dataset to check pipeline function&#39;</span><span class="w"></span>

<span class="w">    </span><span class="c1">// Limit resources so that this can run on GitHub Actions</span><span class="w"></span>
<span class="w">    </span><span class="n">max_cpus</span><span class="w">   </span><span class="o">=</span><span class="w"> </span><span class="mi">2</span><span class="w"></span>
<span class="w">    </span><span class="n">max_memory</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s1">&#39;6.GB&#39;</span><span class="w"></span>
<span class="w">    </span><span class="n">max_time</span><span class="w">   </span><span class="o">=</span><span class="w"> </span><span class="s1">&#39;6.h&#39;</span><span class="w"></span>

<span class="w">    </span><span class="c1">// Input data</span><span class="w"></span>
<span class="w">    </span><span class="n">input</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s1">&#39;https://raw.githubusercontent.com/nf-core/test-datasets/rnaseq/samplesheet/v3.4/samplesheet_test.csv&#39;</span><span class="w"></span>

<span class="w">    </span><span class="c1">// Genome references</span><span class="w"></span>
<span class="w">    </span><span class="n">fasta</span><span class="w">              </span><span class="o">=</span><span class="w"> </span><span class="s1">&#39;https://github.com/nf-core/test-datasets/raw/rnaseq/reference/genome.fasta&#39;</span><span class="w"></span>
<span class="w">    </span><span class="n">gtf</span><span class="w">                </span><span class="o">=</span><span class="w"> </span><span class="s1">&#39;https://github.com/nf-core/test-datasets/raw/rnaseq/reference/genes.gtf.gz&#39;</span><span class="w"></span>
<span class="w">    </span><span class="n">gff</span><span class="w">                </span><span class="o">=</span><span class="w"> </span><span class="s1">&#39;https://github.com/nf-core/test-datasets/raw/rnaseq/reference/genes.gff.gz&#39;</span><span class="w"></span>
<span class="w">    </span><span class="n">transcript_fasta</span><span class="w">   </span><span class="o">=</span><span class="w"> </span><span class="s1">&#39;https://github.com/nf-core/test-datasets/raw/rnaseq/reference/transcriptome.fasta&#39;</span><span class="w"></span>
<span class="w">    </span><span class="n">additional_fasta</span><span class="w">   </span><span class="o">=</span><span class="w"> </span><span class="s1">&#39;https://github.com/nf-core/test-datasets/raw/rnaseq/reference/gfp.fa.gz&#39;</span><span class="w"></span>

<span class="w">    </span><span class="n">bbsplit_fasta_list</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s1">&#39;https://github.com/nf-core/test-datasets/raw/rnaseq/reference/bbsplit_fasta_list.txt&#39;</span><span class="w"></span>
<span class="w">    </span><span class="n">hisat2_index</span><span class="w">       </span><span class="o">=</span><span class="w"> </span><span class="s1">&#39;https://github.com/nf-core/test-datasets/raw/rnaseq/reference/hisat2.tar.gz&#39;</span><span class="w"></span>
<span class="w">    </span><span class="n">salmon_index</span><span class="w">       </span><span class="o">=</span><span class="w"> </span><span class="s1">&#39;https://github.com/nf-core/test-datasets/raw/rnaseq/reference/salmon.tar.gz&#39;</span><span class="w"></span>
<span class="w">    </span><span class="n">rsem_index</span><span class="w">         </span><span class="o">=</span><span class="w"> </span><span class="s1">&#39;https://github.com/nf-core/test-datasets/raw/rnaseq/reference/rsem.tar.gz&#39;</span><span class="w"></span>

<span class="w">    </span><span class="c1">// Other parameters</span><span class="w"></span>
<span class="w">    </span><span class="n">skip_bbsplit</span><span class="w">        </span><span class="o">=</span><span class="w"> </span><span class="kc">false</span><span class="w"></span>
<span class="w">    </span><span class="n">pseudo_aligner</span><span class="w">      </span><span class="o">=</span><span class="w"> </span><span class="s1">&#39;salmon&#39;</span><span class="w"></span>
<span class="w">    </span><span class="n">umitools_bc_pattern</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s1">&#39;NNNN&#39;</span><span class="w"></span>
<span class="o">}</span><span class="w"></span>

<span class="c1">// When using RSEM, remove warning from STAR whilst building tiny indices</span><span class="w"></span>
<span class="n">process</span><span class="w"> </span><span class="o">{</span><span class="w"></span>
<span class="w">    </span><span class="nl">withName:</span><span class="w"> </span><span class="s1">&#39;RSEM_PREPAREREFERENCE_GENOME&#39;</span><span class="w"> </span><span class="o">{</span><span class="w"></span>
<span class="w">        </span><span class="n">ext</span><span class="o">.</span><span class="na">args2</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;--genomeSAindexNbases 7&quot;</span><span class="w"></span>
<span class="w">    </span><span class="o">}</span><span class="w"></span>
<span class="o">}</span><span class="w"></span>
</pre></div>
</div>
<div class="admonition tip">
<p class="admonition-title">Tip</p>
<p>You can find the current version of the above config above <a class="reference external" href="https://github.com/nf-core/rnaseq/blob/master/conf/test.config">here</a></p>
</div>
</section>
<section id="downloading-pipelines">
<h2>Downloading pipelines<a class="headerlink" href="#downloading-pipelines" title="Permalink to this heading"></a></h2>
<p>If your HPC system or server does not have an internet connection you can still run nf-core pipelines by fetching the
pipeline files first and then, manually transferring them to your system.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">nf-core</span> <span class="pre">download</span></code> option simplifies this process and ensures the correct versioning of all the code and containers
needed to run the pipeline. By default, the command will download the pipeline code and the institutional <a class="reference external" href="https://github.com/nf-core/configs">nf-core/configs</a>
files. Again, the <code class="docutils literal notranslate"><span class="pre">-r</span></code> flag allows to fetch a given revision of the pipeline.</p>
<p>Finally, you can also download any singularity image files required by the pipeline, if you specify the <code class="docutils literal notranslate"><span class="pre">--singularity</span></code> flag.</p>
<div class="admonition tip">
<p class="admonition-title">Tip</p>
<p>If you don’t provide any option to <code class="docutils literal notranslate"><span class="pre">`nf-core</span> <span class="pre">download</span></code> an interactive prompt will ask you for the required options.</p>
</div>
<p>We can now try to download the rnaseq pipeline using the command below:</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">nf-core download rnaseq</span>
</pre></div>
</div>
<p>Now we can inspect the structure of the downloaded directory:</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>tree -L <span class="m">2</span> nf-core-rnaseq-3.8.1/

<span class="go">nf-core-rnaseq-3.8.1/</span>
<span class="go">├── configs</span>
<span class="go">│   ├── ..truncated..</span>
<span class="go">│   ├── nextflow.config</span>
<span class="go">│   ├── nfcore_custom.config</span>
<span class="go">│   └── pipeline</span>
<span class="go">├── singularity-images</span>
<span class="go">│   ├── depot.galaxyproject.org-singularity-bbmap-38.93--he522d1c_0.img</span>
<span class="go">│   ├── ..truncated..</span>
<span class="go">│   └── depot.galaxyproject.org-singularity-umi_tools-1.1.2--py38h4a8c8d9_0.img</span>
<span class="go">└── workflow</span>
<span class="go">        ├── CHANGELOG.md</span>
<span class="go">        ├── ..truncated..</span>
<span class="go">        ├── main.nf</span>
<span class="go">        ├── modules</span>
<span class="go">        └── workflows</span>
</pre></div>
</div>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="nextflow_3.html" class="btn btn-neutral float-left" title="Nextflow 3" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="nf-core_2.html" class="btn btn-neutral float-right" title="nf-core for developers" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright .</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>